File: alg-go/Solution373.java
import java.util.*;

class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        // build a sum asc priority queue
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] + a[1] - b[0] - b[1]);
        for (int i = 0; i < nums1.length; i++) {
            pq.offer(new int[]{nums1[i], nums2[0], 0});
        }
        // collect result
        List<List<Integer>> result = new ArrayList();
        while (k-- > 0 && !pq.isEmpty()) {
            int[] curr = pq.poll();
            int i = curr[2];
            result.add(Arrays.asList(curr[0], curr[1]));
            if (i == nums2.length - 1) {
                continue;
            }
            pq.offer(new int[]{curr[0], nums2[i + 1], i + 1});
        }
        return result;
    }
}

File: alg-go/Solution373.go
package main

import (
	"container/heap"
)

type pair struct {
	Sum   int
	Index int
	Nums1 int
	Nums2 int
}

type PriorityQueue []*pair

func (pq PriorityQueue) Len() int {
	return len(pq)
}

func (pq PriorityQueue) Less(i, j int) bool {
    return pq[i].Sum < pq[j].Sum
}

func (pq PriorityQueue) Swap(i, j int) {
    pq[i], pq[j] = pq[j], pq[i]
}

func (pq *PriorityQueue) Push(x interface{}) {
    *pq = append(*pq, x.(*pair))
}

func (pq *PriorityQueue) Pop() interface{} {
    old := *pq
    n := len(old)
    item := old[n-1]
    *pq = old[0 : n-1]
    return item
}

func kSmallestPairs(nums1 []int, nums2 []int, k int) [][]int {

}
